### ISSUE TITLE
Conflict between LLVM's version file and our VERSION file

### ISSUE THREAD

**Issue by matthiasdiener** [UIUC-PPL/charm#2844]
On MacOS, which generally has case-insensitive file systems, the `VERSION` file that gets generated while building Charm++ conflicts with a `version` header file provided by LLVM, leading to compilation errors:

```
clang++ -fPIC -dynamic -fno-common -stdlib=libc++ -DCMK_GFORTRAN -mmacosx-version-min=10.7 -D_DARWIN_C_SOURCE -I../include -I. -c test.cpp -o test.o -std=c++11
In file included from test.cpp:3:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/map:479:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/__tree:14:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/iterator:421:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/__functional_base:14:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/type_traits:417:
In file included from /usr/local/opt/llvm/bin/../include/c++/v1/cstddef:37:
../include/version:1:1: error: expected unqualified-id
61001
^
```

See https://github.com/llvm-mirror/libcxx/blob/master/include/cstddef#L37

This is with llvm 10.0 (homebrew) on MacOS 10.15.4, both old and new build systems are affected.
llvm 9 is probably affected too.

Potential fixes:
- rename our `VERSION` file.
- make our `VERSION` file C-compliant (e.g. by adding `//` before the version number). This would avoid the compilation error, but LLVM's version file would still not be included (which is not a critical error I think).
- remove `VERSION` completely.

**Comment by evan-charmworks** [comment]
Why is there a copy of VERSION in the `include` folder at all?

**Comment by matthiasdiener** [comment]
> Why is there a copy of VERSION in the `include` folder at all?

`include/VERSION` is created when building Charm++ and then symlinked from the main Charm++ directory. The only real use of this file that I could find is inside `charmc` (just to print the Charm++ version number).

**Comment by evan-charmworks** [comment]
What I mean is, why is this file located in `include` instead of a different location that won't get picked up by the preprocessor?

**Comment by matthiasdiener** [comment]
Seems that in the old build system, there are two VERSION files created, with different content:

```
$ cat tmp/VERSION
v6.11.0-devel-254-gf3e432a8e
$ cat include/VERSION
61001
```

This is absolute madness.

**Comment by nilsleiffischer** [comment]
As mentioned in #3044, this issue leads to conflicts with Boost@1.73+, which tries to `#include <version>`. I think renaming the `VERSION` file, moving it to a more charm-specific location such as `include/charm/version.h` or `include/charm_version.h` and/or making it C-compliant would be very useful. We are currently using the file to check the charm version in our CMake build system.

**Comment by matthiasdiener** [comment]
Imho, we should create a C file , `include/charm_version.h`, that merges the content currently split into `tmp/VERSION` and `include/VERSION`:

```c
#define CHARM_VERSION     "v6.11.0-devel-254-gf3e432a8e" 
#define CHARM_API_VERSION 61001 // Must be an int I think
```

We could also add more things there, like `CHARM_API_VERSION_MAJOR` etc.

**Comment by nilsleiffischer** [comment]
@matthiasdiener sounds great!

**Comment by nilsleiffischer** [comment]
@matthiasdiener Please let me know if I can help with that by opening a PR.

**Comment by nilsleiffischer** [comment]
@matthiasdiener do you think this could make it into charm@6.11?

**Comment by matthiasdiener** [comment]
> @matthiasdiener do you think this could make it into charm@6.11?

Could you check if #3112 fixes this issue for you?

**Comment by nilsleiffischer** [comment]
I can confirm that #3112 fixes the issue with Boost v1.74.0 that I raised in #3044

**Comment by BJWiley233** [comment]
The nightly build of NAMD still have version 6.10.2 as well as this issue.  Any reason why version 7.0 is not included in these downloads?

https://www.ks.uiuc.edu/Development/Download/download.cgi?PackageName=NAMD
